doc/source/rllib.rst

```diff
@@ -98,98 +106,137 @@ set ``--checkpoint-freq`` (number of training iterations between checkpoints)
-`the repository <https://github.com/ray-project/ray/blob/master/python/ray/rllib/test/tuned_examples.sh>`__
+`the repository <https://github.com/ray-project/ray/blob/master/python/ray/rllib/tuned_examples>`__
 (some of them are tuned to run on GPUs). If you find better settings or tune
 an algorithm on a different domain, consider submitting a Pull Request!
 
 Python User API
 ---------------
 
-You will be using this part of the API if you run the existing algorithms
-on a new problem. Here is an example how to use it:
+The Python API provides the needed flexibility for applying RLlib to new problems. You will need to use this API if you wish to use custom environments, preprocesors, or models with RLlib.
+
+Here is an example of the basic usage:
 
 ::
 
     import ray
     import ray.rllib.ppo as ppo
 
     ray.init()
-
     config = ppo.DEFAULT_CONFIG.copy()
-    alg = ppo.PPOAgent(config=config, env="CartPole-v1")
+    agent = ppo.PPOAgent(config=config, env="CartPole-v0")
 
-    # Can optionally call alg.restore(path) to load a checkpoint.
+    # Can optionally call agent.restore(path) to load a checkpoint.
 
-    for i in range(10):
-       # Perform one iteration of the algorithm.
-       result = alg.train()
+    for i in range(1000):
+       # Perform one iteration of training the policy with PPO
+       result = agent.train()
        print("result: {}".format(result))
-       print("checkpoint saved at path: {}".format(alg.save()))
+
+       if i % 100 == 0:
+           checkpoint = agent.save()
+           print("checkpoint saved at", checkpoint)
+
+Components: User-customizable and Internal
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following diagram provides a conceptual overview of data flow between different components in RLlib. We start with an ``Environment``, which given an action produces an observation. The observation is preprocessed by a ``Preprocessor`` and ``Filter`` (e.g. for running mean normalization) before being sent to a neural network ``Model``. The model output is in turn interpreted by an ``ActionDistribution`` to determine the next action.
+
+.. image:: rllib-components.svg
+
+The components highlighted in green above are *User-customizable*, which means RLlib provides APIs for swapping in user-defined implementations, as described in the next sections. The purple components are *RLlib internal*, which means they currently can only be modified by changing the RLlib source code.
+
+For more information about these components, also see the `RLlib Developer Guide <rllib-dev.html>`__.
 
 Custom Environments
 ~~~~~~~~~~~~~~~~~~~
 
 To train against a custom environment, i.e. one not in the gym catalog, you
-can register a function that creates the env to refer to it by name. For example:
+can register a function that creates the env to refer to it by name. The contents of the
+``env_config`` agent config field will be passed to that function to allow the
+environment to be configured. For example:
```

Comment:
I think we should specify the interface that the env must implement and provide a skeleton

See #1440.

Reply:
Yeah this is just the Gym env, but it's a bit hard to find the actual interface so let me dig up a good reference.

Reply:
Will this work? https://github.com/openai/gym/blob/master/gym/envs/README.md